lib/image_optim.rb

```diff
@@ -93,15 +104,22 @@ def workers_for_image(path)
   def optimize_image(original)
     original = ImagePath.convert(original)
     return unless (workers = workers_for_image(original))
-    result = Handler.for(original) do |handler|
-      workers.each do |worker|
-        handler.process do |src, dst|
-          worker.optimize(src, dst)
+
+    optimized = @cache.read(original)
+    if optimized.nil?
+      optimized = Handler.for(original) do |handler|
+        workers.each do |worker|
+          handler.process do |src, dst|
+            worker.optimize(src, dst)
+          end
         end
       end
+      @cache.write(original, optimized)
     end
-    return unless result
-    ImagePath::Optimized.new(result, original)
+
+    return unless optimized
+    return unless File.file?(optimized) && !File.symlink?(optimized)
```

Comment:
This should be handled inside `Cache` class. I can propose to combine `Cache#read` and `Cache#write` to a method accepting block, so if there is cache hit, it will be returned, if there is a symlink, nil will be returned, otherwise the block will run and result will be written in cache. So the call would be like:
```ruby
optimized = @cache.fetch(original) do
  Handler.for(original) do |handler|
    ...
  end
end
```